### PR TITLE
Regression suite fixes and added test cases

### LINKED ISSUE #57: Add integration tests into the CI build process for catching regression issues
Add basic integration tests into the CI build process for catching regression issues. Please ensure:

1. Basic install on a fresh machine or docker sandbox takes place as part of the sanity testing. 
2. All API endpoints are verified for regression. It can be as simple as running a Postman script with all the endpoints and testing that models get deployed and inference is working for the examples bundled.

### PR DESCRIPTION
**Pull request by shivamshriwas**
## Description

This PR address following :
 * Adds inference test cases for all models present in model zoo
   -  Text Classification
   - DenseNet
   - SqueezeNet
   - Alexnet
   - Resnet-18
   - VGG11
   - MNIST
   - FAST RCNN
   - MASK RCNN
   - FCN Resnet 101
   - Resnet 152
 * Fixes issues related to executing regression test suite in docker image
 * Fixes issues related to executing regression test suite on EC2 instance
 * Updates corresponding documentation
 * Adds Resnet152 in model zoo.
 * Adds https test cases 
   - get inference/management api description
   - register model
   - scale model sync/async
   - unregister model
 * Fixes Management api collection failures on staging_0_1_1 branch
 * Fixed Pytest failure  on staging_0_1_1 branch

***Commit [17de679](https://github.com/pytorch/serve/commit/17de67901dc170be7989d6daa1fb8b1bf7ab8c8d) from @shivamshriwas has further simplified the task of adding inference related test cases in future.

Fixes #57 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Feature/Issue validation/testing

Executed steps as per https://github.com/pytorch/serve/blob/4fe5ae9aaa5a2b1ffdd5861297c23487bee05fb0/test/README.md

And successfully generated the updated inference and management api reports.

Logs from run on EC2 instance :
[test_exec.log](https://github.com/pytorch/serve/files/4623582/test_exec.log)

Logs from run on docker container :
[test_exec.log](https://github.com/pytorch/serve/files/4623625/test_exec.log)

## Checklist:

- [x] Have you added tests that prove your fix is effective or that this feature works?
- [x] New and existing unit tests pass locally with these changes?
- [x] Have you made corresponding changes to the documentation?

### Note 
The test suite  in this PR can be run on staging_0_1_1  or regression_suite_issue57 branch due to dependency on staging_0_1_1 branch PR# https://github.com/pytorch/serve/pull/322 .

Use this cmd to run test suite.
```bash
./test/regression_tests.sh regression_suite_issue57
```

#### Updated logs
Logs from run on docker container :
[docker_test_exec_log.txt](https://github.com/pytorch/serve/files/4634971/docker_test_exec_log.txt)
Logs from run on EC2 instance :
[ec2_test_exec_log.txt](https://github.com/pytorch/serve/files/4634973/ec2_test_exec_log.txt)

### CODE REVIEW

**Comment by dhanainme** [comment]
Working as expected. Was able to pull the code to my local machine & test it. 

There was a minor issue with respect to sudo usage. Fixed it in my latest commit. 

Test Logs
-----

torchserve_regression_inference

→ Inference Endpoint - Ping
  GET http://localhost:8080/ping [200 OK, 292B, 37ms]

  prepare   wait   dns-lookup   tcp-handshake   transfer-start   download   process   total
  21ms      5ms    161µs        630µs           22ms             7ms        316µs     58ms

  ✓  Status code is 200

→ Model Zoo - Register Text Classification
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/my_text_classifier.mar&model_name=my_text_classifier&initial_workers=1&synchronous=true [200 OK, 299B, 15.6s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       727µs   48µs         155µs           15.6s            2ms        63µs      15.6s

  ✓  Successful POST request

→ Model Zoo - Inference Text Classification
  POST http://localhost:8080/predictions/my_text_classifier [200 OK, 239B, 79ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  4ms       674µs   (cache)      (cache)         76ms             1ms        77µs      83ms

  ✓  Status code is 200

→ Model Zoo - Register Model - Densenet161
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/densenet161.mar&model_name=densenet161&initial_workers=1&synchronous=true [200 OK, 299B, 9.1s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       512µs   (cache)      (cache)         9s               3ms        75µs      9.1s

  ✓  Successful POST request

→ Model Zoo - Inference - DenseNet
  POST http://localhost:8080/predictions/densenet161 [200 OK, 465B, 281ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  2ms       513µs   (cache)      (cache)         277ms            1ms        65µs      282ms

  ✓  Status code is 200

→ Model Zoo - Register Model - SqueezeNet
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/squeezenet1_1.mar&model_name=squeezenet1_1&initial_workers=1&synchronous=true [200 OK, 299B, 1011ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       383µs   (cache)      (cache)         1007ms           2ms        78µs      1011ms

  ✓  Successful POST request

→ Model Zoo - Inference - SqueezeNet
  POST http://localhost:8080/predictions/squeezenet1_1 [200 OK, 459B, 53ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  2ms       316µs   (cache)      (cache)         51ms             1ms        53µs      55ms

  ✓  Status code is 200

→ Model Zoo - Register Model - Alexnet
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/alexnet.mar&model_name=alexnet&initial_workers=1&synchronous=true [200 OK, 299B, 16s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       332µs   (cache)      (cache)         16s              1ms        55µs      16s

  ✓  Successful POST request

→ Model Zoo - Inference - Alexnet
  POST http://localhost:8080/predictions/alexnet [200 OK, 456B, 36ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       282µs   (cache)      (cache)         33ms             1ms        47µs      36ms

  ✓  Status code is 200

→ Model Zoo - Register Model - Resnet-18
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/resnet-18.mar&model_name=resnet-18&initial_workers=1&synchronous=true [200 OK, 299B, 3.7s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       282µs   (cache)      (cache)         3.7s             2ms        50µs      3.7s

  ✓  Successful POST request

→ Model Zoo - Inference - Resnet-18
  POST http://localhost:8080/predictions/resnet-18 [200 OK, 458B, 97ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  2ms       298µs   (cache)      (cache)         94ms             1ms        49µs      98ms

  ✓  Status code is 200

→ Model Zoo - Register Model - VGG11
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/vgg11.mar&model_name=vgg11&initial_workers=1&synchronous=true [200 OK, 299B, 33s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       268µs   (cache)      (cache)         33s              1ms        58µs      33s

  ✓  Successful POST request

→ Model Zoo - Inference - VGG11
  POST http://localhost:8080/predictions/vgg11 [200 OK, 457B, 102ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       256µs   (cache)      (cache)         99ms             1ms        53µs      103ms

  ✓  Status code is 200

→ Model Zoo - Register Model - MNIST
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/mnist.mar&model_name=mnist&initial_workers=1&synchronous=true [200 OK, 299B, 1775ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       269µs   (cache)      (cache)         1772ms           2ms        45µs      1775ms

  ✓  Successful POST request

→ Model Zoo - Inference - MNIST
  POST http://localhost:8080/predictions/mnist [200 OK, 234B, 12ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       239µs   (cache)      (cache)         9ms              1ms        37µs      12ms

  ✓  Status code is 200

→ Model Zoo - Register Model - FAST RCNN
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/fastrcnn.mar&model_name=fastrcnn&initial_workers=1&synchronous=true [200 OK, 299B, 12.5s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       255µs   (cache)      (cache)         12.5s            1ms        43µs      12.5s

  ✓  Successful POST request

→ Model Zoo - Register Model - MASK RCNN
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/maskrcnn.mar&model_name=maskrcnn&initial_workers=1&synchronous=true [200 OK, 299B, 12.3s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       247µs   (cache)      (cache)         12.3s            1ms        44µs      12.3s

  ✓  Successful POST request

→ Model Zoo - Register Model - FCN Resnet 101
  POST http://localhost:8081/models?url=https://torchserve.s3.amazonaws.com/mar_files/fcn_resnet_101.mar&model_name=fcn_resnet_101&initial_workers=1&synchronous=true [200 OK, 299B, 21.1s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       246µs   (cache)      (cache)         21.1s            1ms        38µs      21.1s

  ✓  Successful POST request

→ Model Zoo - Inference - FAST RCNN
  POST http://localhost:8080/predictions/fastrcnn [200 OK, 1.36KB, 1938ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       252µs   (cache)      (cache)         1935ms           1ms        43µs      1939ms

  ✓  Status code is 200

→ Model Zoo - Inference - MASK RCNN
  POST http://localhost:8080/predictions/maskrcnn [200 OK, 1.43KB, 2.1s]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  1ms       250µs   (cache)      (cache)         2.1s             2ms        41µs      2.1s

  ✓  Status code is 200

→ Model Zoo - Inference -FCN Resnet 101
  POST http://localhost:8080/predictions/fcn_resnet_101 [200 OK, 3.17KB, 538ms]

  prepare   wait    dns-lookup   tcp-handshake   transfer-start   download   process   total
  22ms      302µs   (cache)      (cache)         534ms            2ms        46µs      560ms

  ✓  Status code is 200

┌─────────────────────────┬─────────────────────┬────────────────────┐
│                         │            executed │             failed │
├─────────────────────────┼─────────────────────┼────────────────────┤
│              iterations │                   1 │                  0 │
├─────────────────────────┼─────────────────────┼────────────────────┤
│                requests │                  21 │                  0 │
├─────────────────────────┼─────────────────────┼────────────────────┤
│            test-scripts │                  42 │                  0 │
├─────────────────────────┼─────────────────────┼────────────────────┤
│      prerequest-scripts │                  21 │                  0 │
├─────────────────────────┼─────────────────────┼────────────────────┤
│              assertions │                  21 │                  0 │
├─────────────────────────┴─────────────────────┴────────────────────┤
│ total run duration: 2m 12s                                         │
├────────────────────────────────────────────────────────────────────┤
│ total data received: 6.72KB (approx)                               │
├────────────────────────────────────────────────────────────────────┤
│ average response time: 6.3s [min: 12ms, max: 33s, s.d.: 8.8s]      │
├────────────────────────────────────────────────────────────────────┤
│ average DNS lookup time: 105µs [min: 48µs, max: 161µs, s.d.: 56µs] │
├────────────────────────────────────────────────────────────────────┤
│ average first byte time: 6.3s [min: 9ms, max: 33s, s.d.: 8.8s]     │
└────────────────────────────────────────────────────────────────────┘

File: test/README.md [hunk 1]
```diff
@@ -21,14 +21,14 @@ cd serve
 This would build a docker Image with a tag torchserve:1.0 in which we would run our Regression Tests.
```
Comment (chauhang) [inline test/README.md:21]: @shivamshriwas Please update in the instructions as well (:latest tag)

File: test/regression_tests.sh [hunk 2]
```diff
@@ -15,9 +15,17 @@ TEST_EXECUTION_LOG_FILE="/tmp/test_exec.log"
...
-  sudo apt-get install npm
+  pip install  torch torchtext torchvision sentencepiece psutil future
+  pip install  mock pytest pylint pytest-mock pytest-cov
+  apt-get -y install nodejs-dev node-gyp libssl1.0-dev npm
+  npm install -g n
+  n latest
+  export PATH="$PATH"
   npm install -g newman newman-reporter-html
 
   # Clone & Build TorchServe
```
Comment (chauhang) [inline test/regression_tests.sh:31]: @shivamshriwas @mycpuorg We should use the new install from src script instead as part of PR# https://github.com/pytorch/serve/pull/322

**Comment by shivamshriwas** [comment]
> Working as expected. Was able to pull the code to my local machine & test it.
> 
> There was a minor issue with respect to sudo usage. Fixed it in my latest commit.
> 

@dhanainme reverted these changes as it  breaks execution on ec2 without docker.
Test suite can run directly on ec2 w/o docker as well as in docker env. 
Note the changes in README.md line 31 , this will ensure you don't face any  issue related to sudo usage.

File: test/regression_tests.sh (hunk 2 above)
Comment (shivamshriwas) [reply test/regression_tests.sh:26]: Done

File: test/README.md (hunk 1 above)
Comment (shivamshriwas) [reply test/README.md]: Done

File: test/regression_tests.sh (hunk 2 above)
Comment (shivamshriwas) [reply test/regression_tests.sh:26]: We have added test cases for all examples in model zoo which cover both default and custom handler (eg resnet152 and mnist model).

added https test cases covering
- get inference/management api description
- register model
- scale model sync/async
- unregister model

Comment (lokeshgupta1975) [reply test/regression_tests.sh:26]: @chauhang Can you confirm if all the changes that you had requested have been taken care of?

File: test/postman/inference_api_test_collection.json [hunk 3]
```diff
@@ -1,17 +1,17 @@
 {
```
Comment (dhaniram-kshirsagar) [inline test/postman/inference_api_test_collection.json:1]: Good, further templating for data has made it generic and simple + adding new test cases will be adding row in data file

**Comment by dhanainme** [comment]
✅ Test Cases Running as expected in Code Build : https://torchserve-regression-test.s3.amazonaws.com/public/latest/test_exec.log

✅ Test Execution Script fixes are working with both local setup & Code Build

✅ HTTPS Test Cases work 

Logs : 

```

https_test_collection

→ HTTPS Inference API  Description
  OPTIONS https://localhost:8443 [200 OK, 10.08KB, 66ms]

  prepare   wait   dns-lookup   tcp-handshake   ssl-handshake   transfer-start   download   process   total 
  26ms      7ms    155µs        2ms             30ms            19ms             5ms        346µs     92ms  

  ✓  Status code is 200

→ HTTPS Management API Description
  OPTIONS https://localhost:8443 [200 OK, 10.08KB, 8ms]

  prepare   wait    dns-lookup   tcp-handshake   ssl-handshake   transfer-start   download   process   total 
  1ms       637µs   (cache)      (cache)         (cache)         4ms              1ms        81µs      9ms   

  ✓  Status code is 200

→ HTTPS Register Model - SqueezeNet
  POST https://localhost:8444/models?url=https://torchserve.s3.amazonaws.com/mar_files/squeezenet1_1.mar&model_name=squeezenet1_1&initial_workers=1&synchronous=true [200 OK, 355B, 1010ms]

  prepare   wait    dns-lookup   tcp-handshake   ssl-handshake   transfer-start   download   process   total  
  2ms       922µs   28µs         263µs           19ms            986ms            2ms        84µs      1011ms 

  ✓  Successful POST request

→ HTTPS Get SqueezeNet Model Description
  GET https://localhost:8444/models/squeezenet1_1 [200 OK, 750B, 7ms]

  prepare   wait    dns-lookup   tcp-handshake   ssl-handshake   transfer-start   download   process   total 
  1ms       342µs   (cache)      (cache)         (cache)         3ms              1ms        69µs      7ms   

  ✓  Successful GET request

→ HTTPS Scale up Workers -  Synchronous
  PUT https://localhost:8444/models/squeezenet1_1?min_worker=5&max_worker=5&synchronous=true [200 OK, 329B, 791ms]

  prepare   wait    dns-lookup   tcp-handshake   ssl-handshake   transfer-start   download   process   total 
  1ms       376µs   (cache)      (cache)         (cache)         788ms            1ms        64µs      791ms 

  ✓  Successful PUT request

→ HTTPS Scale up Workers - Asynchronous
  PUT https://localhost:8444/models/squeezenet1_1?min_worker=6&max_worker=6&synchronous=false [202 Accepted, 319B, 5ms]

  prepare   wait    dns-lookup   tcp-handshake   ssl-handshake   transfer-start   download   process   total 
  1ms       329µs   (cache)      (cache)         (cache)         2ms              1ms        49µs      6ms   

  ✓  Successful PUT request

→ HTTPS - Inference - SqueezeNet
  POST https://localhost:8443/predictions/squeezenet1_1 [200 OK, 459B, 193ms]

  prepare   wait    dns-lookup   tcp-handshake   ssl-handshake   transfer-start   download   process   total 
  4ms       844µs   (cache)      (cache)         (cache)         190ms            1ms        63µs      197ms 

  ✓  Status code is 200

→ HTTPS UnRegister Model SqueezeNet
  DELETE https://localhost:8444/models/squeezenet1_1 [200 OK, 321B, 36ms]

  prepare   wait    dns-lookup   tcp-handshake   ssl-handshake   transfer-start   download   process   total 
  1ms       358µs   (cache)      (cache)         (cache)         33ms             1ms        51µs      36ms  

  ✓  Successful DELETE request

┌─────────────────────────┬─────────────────────┬────────────────────┐
│                         │            executed │             failed │
├─────────────────────────┼─────────────────────┼────────────────────┤
│              iterations │                   1 │                  0 │
├─────────────────────────┼─────────────────────┼────────────────────┤
│                requests │                   8 │                  0 │
├─────────────────────────┼─────────────────────┼────────────────────┤
│            test-scripts │                   8 │                  0 │
├─────────────────────────┼─────────────────────┼────────────────────┤
│      prerequest-scripts │                   0 │                  0 │
├─────────────────────────┼─────────────────────┼────────────────────┤
│              assertions │                   8 │                  0 │
├─────────────────────────┴─────────────────────┴────────────────────┤
│ total run duration: 2.3s                                           │
├────────────────────────────────────────────────────────────────────┤
│ total data received: 20.57KB (approx)                              │
├────────────────────────────────────────────────────────────────────┤
│ average response time: 264ms [min: 5ms, max: 1010ms, s.d.: 375ms]  │
├────────────────────────────────────────────────────────────────────┤
│ average DNS lookup time: 113µs [min: 28µs, max: 155µs, s.d.: 59µs] │
├────────────────────────────────────────────────────────────────────┤
│ average first byte time: 253ms [min: 2ms, max: 986ms, s.d.: 373ms] │
└────────────────────────────────────────────────────────────────────┘

```

**Pull request merged**